### DISCUSSION TITLE
Some useful ideas

### DISCUSSION THREAD

**Discussion by venividiviciuss** [svsdval/video2midi#54] [Ideas]
I'm going crazy trying to position each key precisely, key by key, to match the keyboard shown in the video exactly... I'm probably overdoing it and I haven't even tested the result yet. I think I've already spent more than an hour on this, and I haven't finished anything yet

IDEA: Unless I'm unnecessarily complicating the situation, could you implement a system that allows to enter the number of keys as an input? In this way, the keys would already be ready to be moved directly.

IDEA 2: This is more a question of order, not a pressing need: could you implement an alignment system that allows to put all the keys at the same height based on the color? For example, I adjust the first key, but the others are misaligned due to manual precision. By pressing a specific key, they would all align on the same horizontal line.

**Comment by svsdval** [comment]
Hmm, this functionality is already created, use the mouse wheel or arrow keys (left, right, up, down) and shift + arrow keys modifications (shift + left, shift + right, shift + up, shift + down), also you can move all the keys by grabbing the right mouse button on the desired key and dragging the entire keyboard. There is no need for all keys to fit on the screen, those that are outside the boundaries will be ignored.

Regarding IDEA 2, can you give an example? Or make a screen recording.

**Reply by venividiviciuss** [reply]
I’m not sure, but currently, the only movements I can perform are: up and down, which shift the black keys vertically, or left and right, which compress all the keys horizontally.

What would be useful is the ability to select a single key and use the arrow keys to move it pixel by pixel. At the moment, as soon as I touch the arrows, all the keys reset, causing me to lose all the adjustments I’ve made.

Regarding the number of keys, I think it would be better to have a way to set this from the beginning. If I start modifying a few and later realize I need more, I would have to resize all the keys, which would disrupt the entire structure.

So here are my suggestions in a clearer format:

Add a variable to set the number of keys at the start (e.g., 1 to 88).
Enable the ability to move selected keys individually using the arrow keys (left, right, up, down).

P.S: _**It might also be useful to draw undetectable rectangles on the video to improve the tracing. This would be helpful only if the tracing is done on the waterfalls and not directly on the keyboard.**_

![midi](https://github.com/user-attachments/assets/a22d54a9-4fe2-475b-a21a-4d72089a519e)


![Screenshot (154)](https://github.com/user-attachments/assets/1c91d37a-947b-48e4-b15a-4d40baa39ae6)

**Reply by svsdval** [reply]
Checkout last update on master brunch.

Regarding draw undetectable rectangles, I think this is only will impair understanding/perception of the whole picture

**Comment by svsdval** [comment]
Ok, I think I understand what you mean. I'll try to implement it =)

**Comment by svsdval** [comment]
Checkout last update on alpha branch =)

Added two buttons:
 1. Vertical alignment of keys to the selected key
 2. Change keys count 
Also added, hot keys to move selected key by 1 pixel, on key pad 4,6,8,2, and 5 to vertical alignment

**Comment by venividiviciuss** [comment]
Hi bro, thanks for the update. I'm trying to test now: but I can't open any video. I hope you can add some exceptions to handle errors too xd Thanks!

Dependencies are installed of course and the videos are in the same directory of v2m.py *:*

![Screenshot (160)](https://github.com/user-attachments/assets/433b691c-8d71-4b2f-94fe-ba95557b8c2b)

**Reply by svsdval** [reply]
try: python .\v2m.py europa.mp4 or  python3 .\v2m.py europa.mp4

**Reply by venividiviciuss** [reply]
Of course, I forgot to do "python"... xD

**Reply by C0rn3j** [reply]
Hmm, that's unfortunate that Windows can't read the shebang.

Looks like it would be solved if this project gets properly packaged with setuptools in pyproject.toml, then one could simply run `v2m` or `video2midi` and it would be an executable on all OSs.
